library/src/main/java/software/amazon/location/auth/utils/AwsRequestSigner.kt

```diff
@@ -0,0 +1,121 @@
+package software.amazon.location.auth.utils
+
+import okhttp3.Headers
+import okhttp3.Request
+import okio.Buffer
+import java.net.URLEncoder
+import java.util.Locale
+import software.amazon.location.auth.utils.Constants.HEADER_AUTHORIZATION
+
+internal const val SIGNING_ALGORITHM = "AWS4-HMAC-SHA256"
+
+/**
+ * Sign the request with the aws authorization header
+ */
+internal fun Request.signed(accessKeyId: String, accessKey: String, region: String, service: String) =
+    newBuilder()
+        .header(HEADER_AUTHORIZATION, awsAuthorizationHeader(accessKeyId, accessKey, region, service))
+        .build()
+
+internal fun Request.awsAuthorizationHeader(accessKeyId: String, accessKey: String, region: String, service: String) =
+    "$SIGNING_ALGORITHM Credential=$accessKeyId/${credentialScope(
+        region,
+        service
+    )}, SignedHeaders=${signedHeaders()}, Signature=${signature(
+        accessKey,
+        region,
+        service
+    )}"
+
+internal fun Request.signature(accessKey: String, region: String, service: String) =
+    hmacSha256(
+        hmacSha256(
+            hmacSha256(
+                hmacSha256(hmacSha256("AWS4$accessKey", amazonDateHeaderShort()), region),
+                service
+            ), "aws4_request"
+        ),
+        stringToSign(region, service)
+    ).toHexString()
+
+internal fun Request.stringToSign(region: String, service: String) =
+    """
+    |$SIGNING_ALGORITHM
+    |${amazonDateHeader()}
+    |${credentialScope(region, service)}
+    |${hash(canonicalRequest())}
+    """.trimMargin("|")
+
+internal fun Request.canonicalRequest() =
+    """
+    |$method
+    |${canonicalUri()}
+    |${canonicalQueryString()}
+    |${canonicalHeaders()}
+    |
+    |${signedHeaders()}
+    |${bodyDigest()}
+    """.trimMargin("|")
+
+private fun Request.canonicalUri() =
+    url.encodedPath.replace(Regex("/+"), "/")
+
+private fun Request.canonicalQueryString() =
+    url.queryParameterNames.sorted()
+        .takeIf { it.isNotEmpty() }
+        ?.flatMap { name ->
+            url.queryParameterValues(name)
+                .filterNotNull()
+                .sorted()
+                .map { value ->
+                    Pair(name.rfc3986Encode(), value.rfc3986Encode())
+                }
+        }
+        ?.joinToString("&") { (name, value) ->
+            "$name=$value"
+        }
+        ?: ""
+
+private fun Request.canonicalHeaders() = headers.canonicalHeaders()
+
+private fun Request.signedHeaders() =
+    headers.names()
+        .map { it.trim().lowercase(Locale.ENGLISH) }
+        .sorted()
+        .joinToString(";")
+
+private fun Request.bodyDigest() =
+    hash(bodyAsString()).lowercase(Locale.ENGLISH)
+
+private fun Request.amazonDateHeader() =
+    header("x-amz-date")
```

Comment:
shouldnt use the already defined constant here?

Reply:
Done